[Federation] Add federated type registry

marun opened the pull request:
This PR adds a registry for federated types. The goal is to simplify the consumption of federated type configuration by controllers/integration tests/e2e tests/upgrade tests.  Consumers can iterate over ``fedtypes.FederatedTypes()``, as per the usage in the crud integration test and controller manager in this PR.

The previous name for the the adapter package - ``typeadapters`` - has been changed to ``fedtypes`` to reflect the fact that more than just type adapters are defined there.  I'm happy to take suggestions on the name, ``fedtypes`` was just the first thing that came to mind. 

cc: @kubernetes/sig-federation-pr-reviews

marun committed:
fed: Add type adapter registration to simplify maintenance

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43713)
<!-- Reviewable:end -->

eparis commented:
/approve
(a one line change in hack)

marun commented:
This approach works for simple federated types, but more complicated types (those involving scheduling or network resources) will require more involved validation than is performed by the current crud tester.  This suggests that the method proposed may need to evolve to allow discovery of the appropriate crud tester for a given adapter type.

perotinus commented:
@marun I like having this as a separate PR: it would be good to get #43500 in and I think this is a separable enough enhancement that it's worth its own PR.

perotinus commented:
@marun Is this ready to review?

marun commented:
Ready for review.

perotinus commented on test/e2e_federation/upgrade.go:
```diff
@@ -25,9 +25,7 @@ import (
-var upgradeTests = []upgrades.Test{
-	upgrades.SecretUpgradeTest(),
-}
+var upgradeTests = upgrades.GenericUpgradeTests()
```
I'm not sure how I'm feeling about the "Generic" name, but I can't think of a more descriptive moniker offhand.

Could we just remove "Generic" altogether and make this UpgradeTests(), or just Tests()? I'm not sure that Generic is adding any semantic information.

marun replied:
I'm assuming there are going to be more involved upgrade tests in the future - e.g. for ingress - and that ``upgradeTests`` will collect both the generic and non-generic tests.

marun replied:
I guess I see this as a case of YAGNI. If/when we get different kinds of upgrade tests, we can rename this function.  But for now it's just a way to separate creation of upgrade tests from their execution, and I'm not sure it's worth bikeshedding over the name of a function that is only ever intended to be called once.

perotinus commented on test/e2e_federation/upgrades/generic.go:
```diff
@@ -58,8 +59,13 @@ func (t *GenericUpgradeTest) Teardown(f *fedframework.Framework) {
-// and after a controlplane upgrade
-func SecretUpgradeTest() Test {
-	return &GenericUpgradeTest{adapter: &typeadapters.SecretAdapter{}}
+func GenericUpgradeTests() []Test {
```
Docs.

marun replied:
Why are docs required for this?  I get that golang religion is 'doc all the things that are exported!' but this is a test-only function, and only used once.

perotinus replied:
I'm a stickler for comments, for several reasons:
- they make the code look more polished, which increases my confidence in its quality
- when writing them, they force me to consider whether I really understand what the thing I'm describing is/is doing
- when writing them, they force me to consider whether the thing I'm describing needs to be two things (e.g., if my comment is "do this and that", maybe I have two functions smushed into one)
- they give a little more context to someone reading the code for the first time
- they help preserve context that seemed obvious but actually wasn't
- they help suggest better function/type/variable/constant names

Of course, not every comment actually accomplishes all of these ends, and there are risks to putting comments on things that don't really need them, but I'm inclined to believe that I'll forget what "GenericUpgradeTests()" means at some point and will want some description.

marun replied:
I like comments, but I think they only have value where intent can't be inferred from code.  Otherwise they're a maintenance burden.  I think for trivial functions such as this - all it does is collect upgrade tests for compatible types - there is minimal benefit from providing a comment.

perotinus commented:
Sorry, I couldn't figure out how to make Reviewable look at only one commit. :/

marun committed:
fed: Update crud integration test to use the the adapter registry

marun committed:
fed: adapter registry -> type registry to enable ctlr mgr use

marun committed:
fed: mv pkg/typeadapters pkg/federatedtypes

marun commented:
@quinton-hoole Feedback appreciated.

perotinus commented:
@marun Can you rebase this? I think it should be good to go from my side, but I'd like to take a look at the final diffs.

shashidharatd commented:
@marun, this pr is still showing 6 commits, the first 2 are already merged, so if you rebase with latest master code this pr should show only the remaining 4 commits. that would simplify the code review as we cannot review per commit in reviewable

marun commented:
rebased

irfanurrehman commented:
A suggestion and a minor nit. This code is evolving nicely.. :-)

---

Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/pkg/fedtypes/adapter.go, line 17 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KhBTATYmuX7zSPWDOXM:-KhBTATYmuX7zSPWDOXN:bk7x719) ([raw file](https://github.com/kubernetes/kubernetes/blob/f34ad32a52cbd9d7ee8f4ba97a4ed9cf15234e89/federation/pkg/fedtypes/adapter.go#L17)):*
> ```Go
> */
> 
> package fedtypes
> ```

Because fedtypes is a little confusing to me atleast.
- types in k8s are generally attributed to api resources, fedtypes seems like we are dealing with some federation level api resources.
- the file name registry.go also had be confused on the first look (because we are used to seeing that name generally attributed to stuff that goes in some backend store).

I actually liked typeadapters.. :-)
My suggestion would be:
- controllertypes for the pkg name and change registry.go to typeregistry.go

(Just ignore this suggestion if you think it will lead to bike shedding)

---

*[federation/pkg/fedtypes/registry.go, line 29 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KhBogSZ0ae_ANKB-ZGc:-KhBogSZ0ae_ANKB-ZGd:blxvbge) ([raw file](https://github.com/kubernetes/kubernetes/blob/f34ad32a52cbd9d7ee8f4ba97a4ed9cf15234e89/federation/pkg/fedtypes/registry.go#L29)):*
> ```Go
> 	Kind              string
> 	ControllerName    string
> 	RequiredResources []schema.GroupVersionResource
> ```

Why name it plural. Wouldn't it be a specific resource always?

---

*[federation/pkg/fedtypes/registry.go, line 36 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KhBy-rDX95I38kQe3Q6:-KhBy-rDX95I38kQe3Q7:bowdj33) ([raw file](https://github.com/kubernetes/kubernetes/blob/f34ad32a52cbd9d7ee8f4ba97a4ed9cf15234e89/federation/pkg/fedtypes/registry.go#L36)):*
> ```Go
> 
> // RegisterFederatedType ensures that configuration for the given kind will be returned by the FederatedTypes method.
> func RegisterFederatedType(kind, controllerName string, requiredResources []schema.GroupVersionResource, factory AdapterFactory) {
> ```

Same for arg name ```requiredResources``` as previous comment

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-:-KhBxeWIkCySNOjKma60:b-4iharj)*
<!-- Sent from Reviewable.io -->

shashidharatd commented on federation/pkg/typeadapters/secret.go:
```diff
@@ -27,20 +27,22 @@ import (
 	kubeclientset "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
 )
 
+const secretKind = "secret"
```
better to use camel case for secretKind, it is the convention used for kube object kind.

shashidharatd commented on federation/pkg/typeadapters/secret.go:
```diff
@@ -27,10 +27,10 @@ import (
 )
 
-const secretKind = "secret"
+const SecretKind = "secret"
```
better to use camel case for SecretKind, it is the convention used for kube object kind.

shashidharatd commented on federation/pkg/typeadapters/secret.go:
```diff
@@ -27,10 +28,13 @@ import (
 
-const SecretKind = "secret"
+const (
+	SecretKind           = "secret"
```
better to use camel case for SecretKind, it is the convention used for kube object kind.

shashidharatd commented on federation/pkg/fedtypes/secret.go:
```diff
@@ -27,20 +28,25 @@ import (
 )
 
+const (
+	SecretKind           = "secret"
```
better to use camel case for SecretKind, it is the convention used for kube object kind.

perotinus commented on federation/pkg/fedtypes/adapter.go:
```diff
@@ -57,3 +55,9 @@ type FederatedTypeAdapter interface {
 }
+
+// AdapterFactory defines the function signature for factory methods
+// that create FederatedFedtypes.  Such methods should be
```
Should this be FederatedTypeAdapters instead of FederatedFedtypes?

marun commented:
rebased and addressed comments

---

Review status: 0 of 18 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[federation/pkg/fedtypes/adapter.go, line 17 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KhBTATYmuX7zSPWDOXM:-KhcxbO4d7M-HKwXlK6T:b-u5aoa) ([raw file](https://github.com/kubernetes/kubernetes/blob/f34ad32a52cbd9d7ee8f4ba97a4ed9cf15234e89/federation/pkg/fedtypes/adapter.go#L17)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Because fedtypes is a little confusing to me atleast.
- types in k8s are generally attributed to api resources, fedtypes seems like we are dealing with some federation level api resources.
- the file name registry.go also had be confused on the first look (because we are used to seeing that name generally attributed to stuff that goes in some backend store).

I actually liked typeadapters.. :-)
My suggestion would be:
- controllertypes for the pkg name and change registry.go to typeregistry.go

(Just ignore this suggestion if you think it will lead to bike shedding)
</blockquote></details>

In consultation with @perotinus and @csbell I've renamed fedtypes to federatedtypes.  The package is intended to store code, configuration and test helpers for federated types - it's not just for configuring controllers.  I hope that clears up the ambiguity of fedtypes, which as you point out could have meant 'federation' types.

Yes, kube has pkg/api/registry.go, but I think the fact that this isn't in an api package (and doesn't have a corresponding types.go file) is a clear indication that the file does not have the same purpose.

---

*[federation/pkg/fedtypes/adapter.go, line 60 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-Kh5hw6iEDmumybf5JC--r5-60:-KhcyoQAaLUpHBOnn96N:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/f34ad32a52cbd9d7ee8f4ba97a4ed9cf15234e89/federation/pkg/fedtypes/adapter.go#L60)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Should this be FederatedTypeAdapters instead of FederatedFedtypes?
</blockquote></details>

Done.

---

*[federation/pkg/fedtypes/registry.go, line 29 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KhBogSZ0ae_ANKB-ZGc:-Khcyq4Z2ZOiBWII0i1l:b-xlwo39) ([raw file](https://github.com/kubernetes/kubernetes/blob/f34ad32a52cbd9d7ee8f4ba97a4ed9cf15234e89/federation/pkg/fedtypes/registry.go#L29)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Why name it plural. Wouldn't it be a specific resource always?
</blockquote></details>

This is the name of the variable for all existing controllers, and I don't want to change it in this PR.  You are correct that no controller specifies more than a single resource, but that may be in error.  Ingress, for example, relies on configmaps as well as ingresses.  

@nikhiljindal, you recently added ``RequiredResources``  to controllers other than ingress.  What are your thoughts?

---

*[federation/pkg/fedtypes/registry.go, line 36 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KhBy-rDX95I38kQe3Q6:-Khd02LANaSU4i84PWsy:b-7ropln) ([raw file](https://github.com/kubernetes/kubernetes/blob/f34ad32a52cbd9d7ee8f4ba97a4ed9cf15234e89/federation/pkg/fedtypes/registry.go#L36)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Same for arg name ```requiredResources``` as previous comment
</blockquote></details>

ditto

---

*[federation/pkg/fedtypes/secret.go, line 30 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KgFijYYRsDWDDQGWwyp-r2-30:-Khd0G788DGYSWUe3rEa:b-uwnwrn) ([raw file](https://github.com/kubernetes/kubernetes/blob/7e99fec97c384507cb2a8c97aa7e2a9f83963c39/federation/pkg/fedtypes/secret.go#L30)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

better to use camel case for secretKind, it is the convention used for kube object kind.
</blockquote></details>

Can you please remove these duplicated comments?

---

*[federation/pkg/fedtypes/secret.go, line 32 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KgFijYYRsDWDDQGWwyp-r4-32:-Khd-JI1IRQw3wynSLeh:bl9zuze) ([raw file](https://github.com/kubernetes/kubernetes/blob/2f71d7aa0162dbac0ce0ae16168f7b7af0dfad54/federation/pkg/fedtypes/secret.go#L32)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

better to use camel case for SecretKind, it is the convention used for kube object kind.
</blockquote></details>

I've been following the convention used in the existing secret controller.  I agree with you that being consistent with Kube is a good idea, and I'd like to tackle s/secret/Secret separately since it implies changing how the sync controller logs (i.e. rather than %q+kind, %s+kind will likely make more sense). 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-:-KhcxYxZrTXffvvsnwkm:b-nfd7ht)*
<!-- Sent from Reviewable.io -->

marun commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

shashidharatd commented:
Review status: 0 of 18 files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federatedtypes/secret.go, line 32 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713#-KgFijYYRsDWDDQGWwyp-r4-32:-Khdtks41p2ecFdNrXoF:b-1qf1rn) ([raw file](https://github.com/kubernetes/kubernetes/blob/2f71d7aa0162dbac0ce0ae16168f7b7af0dfad54/federation/pkg/federatedtypes/secret.go#L32)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I've been following the convention used in the existing secret controller.  I agree with you that being consistent with Kube is a good idea, and I'd like to tackle s/secret/Secret separately since it implies changing how the sync controller logs (i.e. rather than %q+kind, %s+kind will likely make more sense). 
</blockquote></details>

ok. fine

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43713)*
<!-- Sent from Reviewable.io -->

marun commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

marun commented:
rebased 

@k8s-bot pull-kubernetes-federation-e2e-gce test this

marun commented:
@csbell e2e is green; http://prow.k8s.io/?type=presubmit&pull=43713&job=pull-kubernetes-federation-e2e-gce

csbell commented:
Neat.
/lgtm
